### DISCUSSION TITLE
Split out route params and search params in router.query

### DISCUSSION THREAD

**Discussion by benkingcode** [vercel/next.js#34849] [Ideas]
### Describe the feature you'd like to request

`router.query` currently returns both route params and search params mangled together. For example, a page at `/pages/[slug]` and a URL of `https://example.com/pages/example-slug?foo=bar` would return;

```js
{
  "slug": "example-slug",
  "foo": "bar"
}
```

This is problematic for us as we now can't rely on this object. The search params sent to the origin are different from the request as we sit the Next.js app behind a Cloudflare Worker which strips query params before being sent (in order to improve cache hits).

From the client we can simply use `window.location.search` to get the browser location's search params. However, we can't get an object that consists of _just_ the route params such as `slug`.

### Describe the solution you'd like

In order to fix this in a backwards-compatible way, I would suggest adding a new object to `router` called `routeParams` which is similar to `query`, but only includes the params from the route and does not include the search params.

For example,

```js
{
  ...
  "query": {
    "slug": "example-slug",
    "foo": "bar"
  },
  "routeParams": {
    "slug": "example-slug",
  }
}
```

### Describe alternatives you've considered

Right now we are using this wrapper around `useRouter` which is pretty hacky;

```ts
interface ICacheSafeRouter extends NextRouter {
  query: never
  _serverSideQueryParamsNotCacheSafe: NextRouter['query']
  routeParams: qs.ParsedUrlQuery
  searchParams: qs.ParsedUrlQuery
}

export function useCacheSafeRouter(): ICacheSafeRouter {
  const router = useRouter()

  const routeParams = { ...router.query }
  Object.keys(router.query).forEach((param) => {
    if (!router.pathname.includes(param)) {
      delete routeParams[param]
    }
  })

  return {
    ...router,
    get query(): never {
      throw new Error(
        'Accessing query is not allowed on cacheSafeRouter, use routeParams or searchParams'
      )
    },
    _serverSideQueryParamsNotCacheSafe: router.query,
    routeParams,
    get searchParams() {
      if (typeof window === 'undefined') {
        throw new Error('Accessing searchParams is not allowed server-side')
      }

      return qs.parse(window.location.search.split('?')[1])
    },
  }
}

```

**Comment by michielmetcake** [comment]
I'd love to see this! Because it makes sense

**Comment by bertho-zero** [comment]
Mixing the `query` and the `params` is problematic in my opinion, they don't have the same utility.

I have to separate them to use the `query` (without `params`) and I have to merge them to use the router.
